src/mlpack/methods/preprocess/preprocess_string_main.cpp

```diff
@@ -0,0 +1,273 @@
+    }
+  }
+  return dimensions;
+}
+
+static void WriteOutput(const string& outputFilename,
+                 const vector<vector<string>>& dataset,
+                 const unordered_map<size_t, vector<string>>& nonNumericInput,
+                 const string& columnDelimiter,
+                 const unordered_set<size_t>& dimensions)
+{
+  ofstream fout(outputFilename, ios::trunc);
+  if (!fout.is_open())
+    Log::Fatal << "Unable to open a file for writing output.\n";
+  for (size_t i = 0 ; i < dataset.size(); i++)
+  {
+    for (size_t j = 0 ; j < dataset[i].size(); j++)
+    {
+      if (dimensions.find(j) != dimensions.end())
+      {
+        if (j + 1 < dataset[i].size())
+          fout<<nonNumericInput.at(j)[i]<<columnDelimiter;
+        else
+          fout<<nonNumericInput.at(j)[i];
+      }
+      else
+      {
+        if (j + 1 < dataset[i].size())
+          fout<<dataset[i][j]<<columnDelimiter;
+        else
+          fout<<dataset[i][j];
+      }
+    }
+    fout<<"\n";
+  } 
+}
+
+static void mlpackMain()
+{
+  // Parse command line options.
+  // Extracting the filename
+  const string filename = CLI::GetParam<string>("actual_dataset");
+  // This is very dangerous, Let's add a check tommorrow.
+  string columnDelimiter;
+  if (CLI::HasParam("columnDelimiter"))
+  {
+    columnDelimiter = CLI::GetParam<string>("columnDelimiter");
+    // Allow only 3 delimiters.
+    RequireParamValue<string>("columnDelimiter", [](string del)
+        { return del == "\t" || del == "," || del == " "; }, true,
+        "Delimiter should be either \\t (tab) or , (comma) or ' ' (space) ");
+  }
+  else
+  {
+      if (data::Extension(filename) == "csv")
+      {
+        columnDelimiter = ",";
+        Log::Warn << "Found csv Extension, taking , as columnDelimiter. \n";
+      }
+      else if (data::Extension(filename) == "tsv" ||
+          data::Extension(filename) == "txt")
+      {
+        columnDelimiter = "\t";
+        Log::Warn << "Found tsv or txt Extension, taking \\t as"
+        "columnDelimiter. \n";
+      }
+      else
+      {
+        columnDelimiter = "\t";
+        Log::Warn << "columnDelimiter not specified, taking default value \n";
+      }
+  }
+  // Handling Dimension vector
+  vector<string> tempDimension =
+      CLI::GetParam<vector<string> >("dimension");
+  unordered_set<size_t> dimensions = GetColumnIndices(tempDimension);
+  vector<vector<string>> dataset = CreateDataset(filename, columnDelimiter[0]);
+  for (auto colIndex : dimensions)
+    if (colIndex >= dataset.back().size())
+      Log::Fatal << "The index given is out of range, please verify \n";
+  // Preparing the input dataset on which string manipulation has to be done.
+  // vector<vector<string>> nonNumericInput(dimension.size());
+  unordered_map<size_t , vector<string>> nonNumericInput;
+  for (size_t i = 0; i < dataset.size(); i++)
+  {
+    for (auto& datasetCol : dimensions)
+      nonNumericInput[datasetCol].push_back(dataset[i][datasetCol]);
```

Comment:
I guess you are not going to use `dataset[i][datasetCol]` anymore. In that case it's better to move it to the vector than copy.
```
nonNumericInput[datasetCol].push_back(move(dataset[i][datasetCol]));
```

Reply:
I need it for writting results back to the ouptut file, since preprocessed data is in nonNumericInput and dataset has numeric input.

Reply:
I mean you don't need to store the columns that you move to `nonNumericInput` twice.